rfcs/056-prismic-etl-pipeline/README.md

```diff
@@ -0,0 +1,91 @@
+
+A basic example of this would be that the `article` type has a `contributors` field, which links to `role` and `person` types:
+```json
+{
+  ...
+  "contributors": {
+    "role": {
+      "id": "<foreign key>",
+      "type": "editorial-contributor-roles",
+      ...
+    },
+    "person": {
+      "id": "<foreign key>",
+      "type": "people",
+      ...
+    }
+  }
+}
+```
+
+Fortunately, Prismic provide an API, [`graphQuery`](https://prismic.io/docs/graphquery-rest-api) which can do this denormalisation for us, using a GraphQL-like syntax. This straightforwardly solves the first problem of denormalising linked data, but not the second problem of reflecting changes in it.
+
+The proposed solution to this is to store the IDs of the secondary documents alongside the other information we index in Elasticsearch for primary documents:
+```json
+{
+  "display": {
+    // Opaque JSON to be displayed by the API
+  },
+  "query": {
+    // Fields for querying, filtering etc
+    ...
+    // List of identifiers (foreign keys) of linked documents
+    "ids": [
+      "1",
+      "2",
+      "3"
+    ]
+  }
+}
+```
+Then we can build a pipeline that works as follows:
+
+1. A 'window generator' Lambda triggered on schedule, which generates a payload representing a time period/window that is sent to (2).
+2. A 'Prismic ETL' Lambda which consumes time periods (potentially half-bounded or unbounded, for complete reindexes) and then:
+   1. Queries Prismic for all documents (including denormalised data on primary documents) updated within the time window.
+   2. For all secondary documents, query the ES index for already-indexed primary documents that contain them.
+   3. Queries Prismic for all the documents (including denormalised data) from (ii) that are not part of the list returned by (i).
+   4. Transforms the resultant primary documents into JSON objects as described above.
+   5. Indexes these into an Elasticsearch cluster using the Elasticsearch JS client's bulk helpers. 
+
+For complete reindexes, it would be straightforward to trigger the Prismic ETL lambda from a local script, with a payload that covers all documents. In this case, it would also be an easy optimisation to disable steps (ii) and (iii), as all documents would be being fetched regardless.
+
+#### Technical implementation points
+- The intention is that the Prismic ETL Lambda will be written in TypeScript for maintainability, but one disadvantage of this is that we lose some of the patterns/tools that Scala gives us for reactive streaming data pipelines. Suggest we try [RxJS](https://rxjs.dev/) for this purpose.
```

Comment:
There's a stronger argument for TS: sharing code/dev expertise between the existing front-end apps.

Reply:
Indeed, that's what I meant by maintainability - will clarify.